File: backend/src/models/grade.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import Column, String, DateTime, ForeignKey, Float, Text, Integer
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship

from src.utils.database import Base


class Grade(Base):
    __tablename__ = "grades"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    student_id = Column(UUID(as_uuid=True), ForeignKey("students.id"), nullable=False, index=True)
    teacher_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    exam_name = Column(String(255), nullable=False)
    subject = Column(String(100), nullable=False)
    score = Column(Float, nullable=False)
    max_score = Column(Float, nullable=False, default=100.0)
    wrong_questions = Column(JSONB, nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    student = relationship("Student", back_populates="grades")
    teacher = relationship("User")

    def __repr__(self) -> str:
        return f"<Grade {self.exam_name} - {self.score}/{self.max_score}>"

File: desktop_app/lib/models/grade.dart
class GradeModel {
  final String id;
  final String studentId;
  final String examName;
  final String subject;
  final double score;
  final double maxScore;
  final List<int>? wrongQuestions;
  final String? notes;
  final DateTime createdAt;

  GradeModel({
    required this.id,
    required this.studentId,
    required this.examName,
    required this.subject,
    required this.score,
    required this.maxScore,
    this.wrongQuestions,
    this.notes,
    required this.createdAt,
  });

  factory GradeModel.fromJson(Map<String, dynamic> json) {
    return GradeModel(
      id: json['id'],
      studentId: json['student_id'],
      examName: json['exam_name'],
      subject: json['subject'],
      score: (json['score'] ?? 0).toDouble(),
      maxScore: (json['max_score'] ?? 100).toDouble(),
      wrongQuestions: json['wrong_questions'] != null
          ? List<int>.from(json['wrong_questions'])
          : null,
      notes: json['notes'],
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'student_id': studentId,
        'exam_name': examName,
        'subject': subject,
        'score': score,
        'max_score': maxScore,
        'wrong_questions': wrongQuestions,
        'notes': notes,
        'created_at': createdAt.toIso8601String(),
      };

  double get percentage => maxScore > 0 ? (score / maxScore) * 100 : 0;
}
